Lib9c/Action/DPoS/Control/SlashCtrl.cs

```diff
@@ -0,0 +1,472 @@
+using Libplanet.Action;
+using Libplanet.Action.State;
+using Libplanet.Crypto;
+using Libplanet.Types.Assets;
+using Nekoyume.Module;
+using System;
+using System.Linq;
+using System.Collections.Immutable;
+
+namespace Nekoyume.Action.DPoS.Control
+{
+    internal static class SlashCtrl
+    {
+        public static IWorld Slash(
+            IWorld world,
+            IActionContext actionContext,
+            Address validatorAddress,
+            long infractionHeight,
+            BigInteger power,
+            BigInteger slashFactor,
+            IImmutableSet<Currency> nativeTokens)
+        {
+            if (slashFactor < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(slashFactor), "Slash factor must be greater than or equal to 0.");
+            }
+
+            if (!(ValidatorCtrl.GetValidator(world, validatorAddress) is { } validator))
+            {
+                throw new NullValidatorException(validatorAddress);
+            }
+
+            var amount = FungibleAssetValue.FromRawValue(Asset.ConsensusToken, power);
+            var (slashAmount, r) = amount.DivRem(slashFactor);
+
+            if (validator.Status == BondingStatus.Unbonded)
+            {
+                throw new InvalidOperationException("should not be slashing unbonded validator");
+            }
+
+            var remainingSlashAmount = slashAmount;
+
+            if (infractionHeight > actionContext.BlockIndex)
+            {
+                throw new ArgumentOutOfRangeException(
+                    "impossible attempt to slash future infraction",
+                    nameof(infractionHeight));
+            }
+            else if (infractionHeight == actionContext.BlockIndex)
+            {
+            }
+            else if (infractionHeight < actionContext.BlockIndex)
+            {
+                var unbondingSet = UnbondingSetCtrl.GetUnbondingSet(world)!;
+                foreach (var item in unbondingSet.UndelegationAddressSet)
+                {
+                    var undelegation = UndelegateCtrl.GetUndelegation(world, item)!;
+                    if (undelegation.ValidatorAddress == validatorAddress)
+                    {
+                        world = SlashUnbondingDelegation(world, actionContext, undelegation, infractionHeight, slashFactor, out var amountSlashed);
+                        remainingSlashAmount -= amountSlashed;
+                    }
+                }
+
+                foreach (var item in unbondingSet.RedelegationAddressSet)
+                {
+                    var redelegation = RedelegateCtrl.GetRedelegation(world, item)!;
+                    if (redelegation.SrcValidatorAddress == validatorAddress)
+                    {
+                        world = SlashRedelegation(world, actionContext, redelegation, infractionHeight, slashFactor, nativeTokens, out var amountSlashed);
+                        remainingSlashAmount -= amountSlashed;
+                    }
+                }
```

Comment:
I recommend separating this part into a separate method to make it more semantic.

Reply:
Separated code into the 'SlashRedelegations' method.